Cython: emit `CYTHON_LIMITED_API`

amcn opened the pull request:
If the user has specified the `limited_api` keyword arg on an 'extension_module' invocation, then add `CYTHON_LIMITED_API` to the C and C++ definitions of the target (along with the existing `Py_LIMITED_API`).

Cython's support for this functionality was added in version 3.0, so we issue a warning if the detected Cython compiler is not at least this version.

Two tests have been added: one based on Cython's limited API test which aims to test that it works in the case of the user's compiler supporting this functionality, and another to test that a warning is issued in the opposite case.

eli-schwartz commented on docs/markdown/snippets/cython-limited-api.md:
```diff
@@ -0,0 +1,6 @@
+Using the `limited_api` keyword when declaring a Python `extension_module`
+now causes Meson to define the `CYTHON_LIMITED_API` symbol. Cython's support
+for Python's Limited API was added in Cython version 3. If Meson detects a
+version of Cython earlier than this, a warning is raised.
```
I think we don't need to document in the release notes, under what circumstances we emit an error.

amcn replied:
Ok, I'll remove the release note snippet.

eli-schwartz replied:
Sorry. To be clear, it's a new feature and needs a release notes snippet.

The question is how much detail should the snippet include. Usually we describe the new feature. It's unusual to *also* describe what circumstances emit a warning. The primary documentation is probably a better place for that.

eli-schwartz commented on docs/markdown/Cython.md:
```diff
@@ -60,3 +60,24 @@ py.extension_module(
+
+Cython's support for Python's Limited API was introduced in Cython version 3.0.
+If Meson detects that the user's Cython compiler is not greater than or equal to
+version 3.0, a warning is issued.
```
We *should* note that here, though, yes.


"If the cython version is too old, meson issues a warning but proceeds" may be a better phrasing. The key information here that users care about is that it is safe/allowable to request the limited API without checking the cython version.

At least from a documentation standpoint.

amcn replied:
Thanks, I wasn't sure about the wording here at all. I'll update to your suggestion.

eli-schwartz commented on test cases/warning/10 cython limited api version/meson.build:
```diff
@@ -0,0 +1,12 @@
+
+cythonc = meson.get_compiler('cython')
+if cythonc.version() >= '3.0'
+  error('MESON_SKIP_TEST: Cython compiler supports this functionality, no warning necessary')
```
I'm not sure it's valuable to have a test case testing that we emit a warning. This is running an additional project cycle just to verify the UX, but it's not obvious to me that the cython version check is something that is at risk of being gotten incorrect.

amcn replied:
I'll remove this.

eli-schwartz commented on test cases/cython/4 limited api/test.json:
```diff
@@ -0,0 +1,10 @@
+  "matrix": {
+    "options": {
+      "cython_language": [
+         { "val": "c" },
+         { "val": "cpp" }
```
Aside: is there anything interesting about handling C vs C++ here? I notice you are testing both (so, two tests not one).

amcn replied:
Ah good catch: this is a vestige from a previous attempt at making a valid test. I'll remove this, I don't think there is a difference between C and C++ here.

eli-schwartz reviewed (commented):
There's an interesting question here about what the behavior should be if we warn but continue when cython doesn't support the limited API.

Does the result actually compile correctly if we pass the relevant python defines? If it does -- I would think the abi3 tags would then be wrong.

amcn commented:
> There's an interesting question here about what the behavior should be if we warn but continue when cython doesn't support the limited API.

I wasn't sure about what the behaviour should be, and I'm not at all sure this PR implements anything actually sensible from a Meson user's perspective, so yeah a discussion with Cython stakeholders would be very useful here. 

> Does the result actually compile correctly if we pass the relevant python defines? If it does -- I would think the abi3 tags would then be wrong.

In all the examples I have found thus far, defining `Py_LIMITED_API` for the compilation of any C code generated by Cython without also defining `CYTHON_LIMITED_API` yields many compilation errors, as Cython by default targets the full non-limited Python C API. Is that what you mean by relevant python defines?

eli-schwartz commented:
> In all the examples I have found thus far, defining `Py_LIMITED_API` for the compilation of any C code generated by Cython without also defining `CYTHON_LIMITED_API` yields many compilation errors, as Cython by default targets the full non-limited Python C API. Is that what you mean by relevant python defines?

Yes -- it sounds like we should just make it a fatal meson error.

amcn commented:
> Yes -- it sounds like we should just make it a fatal meson error.

The current logic is if the user has specified a `limited_api` kwarg and Meson finds that a `cython` compiler has been registered, then raise a warning and continue.

If I change the current warning to a fatal error then a meson.build such as this will raise an error:

```
add_languages('c', 'cython')
py.extension_module('my_extension',
  'my_extension.c',
  limited_api: '3.7'
)
```

Even though the extension_module in this case does not use any Cython files as input, Meson will raise an error, which seems wrong to me.

Initially I had implemented the raising of a fatal error back when I was doing #11745 but it ended up being quite messy: trying to work out for a given `extension_module` invocation if the user had supplied both the `limited_api` kwarg and any Cython files as input, then do a check if the Cython compiler version is sufficient and if not, raise a fatal error. Finding out if any of the input files are intended as Cython ended up being the sticking point because they could be generated sources. Is there a way to do this reliably?

eli-schwartz commented:
IIRC `compiler.file_suffixes`, plus getting the filename from any source that isn't `isinstance(sourcefile, str)`.

amcn commented:
> IIRC `compiler.file_suffixes`, plus getting the filename from any source that isn't `isinstance(sourcefile, str)`.

Interesting! I shall implement this, thanks for the feedback!

amcn commented:
Oops, I forgot about this PR. I think I have addressed most of the feedback received so far. I'll set this to ready for review for another round of feedback.

dcbaker commented:
There’s some 1.4 references that need to be 1.5 now

amcn commented:
I've updated the documentation to refer to 1.5 now instead of 1.4.

dcbaker reviewed (approved):
This looks good to me. @eli-schwartz, do you have any more comments?

amcn commented:
Oops, I completely forgot about this PR. I'm going to rebase it and see what needs to be updated.

amcn commented:
> Oops, I completely forgot about this PR. I'm going to rebase it and see what needs to be updated.

The documentation has been updated to refer to 1.6

amcn commented:
I was reading the in-development version of Cython's docs and they [now make mention](https://github.com/cython/cython/commit/10736826646c894b8b2c9cd3799ce7142707c8ed) of meson's support for the Python limited API.

This is currently a bit problematic because the meson example in their docs will not work correctly without this PR being merged. We should probably sync with the Cython devs to make sure everyone is on the same page regarding what is communicated to users about what versions of meson and Cython are needed.

Also worth noting: the Cython documentation now states that while its support for the limited API was initially made available in version 3.0, they don't consider that version's support to be practically useful, instead recommending 3.1. There is an explicit version check for 3.0 in this PR which perhaps in light of this should be changed to 3.1

Tagging @da-woods, @eli-schwartz for visibility.

da-woods commented:
The current master version of Cython only requires `Py_LIMITED_API` to be defined and assumes that if you've set that then you want Cython to adapt. 3.0 would require `CYTHON_LIMITED_API` to be defined.

Setting `CYTHON_LIMITED_API` explicitly remains fine (even if unnecessary) so there's no harm in doing it.

amcn commented:
Given the update from @da-woods and the information that Cython already examines `Py_LIMITED_API`, it would seem that this PR is not necessary.

If nobody objects I will close this PR in a few days in the interest of both repository hygiene and also not introducing code into Meson that serves little purpose beyond complicating maintenance.

amcn commented:
Closing this now as there doesn't seem to be any objections. We can always revisit this if necessary.
